### PR TITLE
[AddressLowering] Support coroutines

### PR DESCRIPTION
**Pull request by meg-gupta**

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/Mandatory/AddressLowering.cpp [hunk 1]
```diff
@@ -2080,6 +2094,34 @@ void ApplyRewriter::rewriteApply(ArrayRef<SILValue> newCallArgs) {
   // will be deleted with its destructure_tuple.
 }
 
+void ApplyRewriter::convertBeginApplyWithGuaranteedYield() {
+  auto *origCall = cast<BeginApplyInst>(apply.getInstruction());
+  SmallVector<SILValue, 4> opValues;
+  auto argOps = origCall->getArgumentOperands();
+  llvm::transform(argOps, std::back_inserter(opValues),
+                  [](Operand &op) { return op.get(); });
```
Comment (atrick) [inline lib/SILOptimizer/Mandatory/AddressLowering.cpp:2102]: I know this is done elsewhere, so it's your choice. But it's a lot easier to read and debug, and faster to build range-based for loops:
`for (auto &oper : getArgumentOperand()) { opValues.push_back(oper.get()); }`

**Review by atrick — APPROVED** [review]
LGTM thanks!

I just scanned the code. So hopefully you can give some thought to coverage of the various combinations of argument conventions...

**Review by atrick — COMMENTED** [review]
@meg-gupta things to check and handle...

- indirect argument conventions with loadable types
- dead results
- loadable yields
- whatever I didn't think of

better to have coverage now than debug later

**Comment by meg-gupta** [comment]
@atrick Thanks for listing them. I have added more test cases now.

File: lib/SILOptimizer/Mandatory/AddressLowering.cpp (hunk 1 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Mandatory/AddressLowering.cpp]: I prefer this version as well. But I think I started using the llvm::transform version after getting feedback a couple of times.

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Comment by meg-gupta** [comment]
@swift-ci smoke test

**Pull request merged**